app/src/main/java/com/example/criminalintent/data/local/CrimeDao.kt

```diff
@@ -18,18 +18,15 @@ interface CrimeDao {
     @Query("SELECT * FROM crime WHERE id =(:id)")
     fun getCrime(id: Int): LiveData<Crime?>
 
-    @Update
+    @Update(onConflict = OnConflictStrategy.IGNORE)
     fun updateCrime(crime: Crime)
 
-    @Insert
+    @Insert(onConflict = OnConflictStrategy.IGNORE)
     fun addCrime(crime: Crime)
 
-    fun upsertCrime(crime: Crime) {
-        try {
-            addCrime(crime)
-        } catch (exception: SQLiteConstraintException) {
-            Log.e("exception", "upsert: ${exception.message}")
-            updateCrime(crime)
-        }
+    @Transaction
+    fun addOrUpdate(crime: Crime) {
+        addCrime(crime)
+        updateCrime(crime)
```

Comment:
You should not update as well if the crime is non existent. As the answer says, use the response from add to know if there is an existing item with this primary key already